cmd/deploy/deploy.go

```diff
@@ -0,0 +1,458 @@
+	log.Infof("Deploying Meshnet from: %s", mPath)
+	if err := execCmd("kubectl", "apply", "-k", mPath); err != nil {
+		return err
+	}
+	log.Infof("Meshnet Deployed")
+	return nil
+}
+
+func (m *MeshnetSpec) Healthy(ctx context.Context) error {
+	log.Infof("Waiting on Metallb to be Healthy")
+	w, err := m.kClient.AppsV1().DaemonSets("meshnet").Watch(ctx, metav1.ListOptions{
+		FieldSelector: fields.SelectorFromSet(fields.Set{metav1.ObjectNameField: "meshnet"}).String(),
+	})
+	if err != nil {
+		return err
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context canceled before healthy")
+		case e, ok := <-w.ResultChan():
+			if !ok {
+				return fmt.Errorf("watch channel closed before healthy")
+			}
+			d, ok := e.Object.(*appsv1.DaemonSet)
+			if !ok {
+				return fmt.Errorf("invalid object type: %T", d)
+			}
+			if d.Status.CurrentNumberScheduled == d.Status.CurrentNumberScheduled &&
+				d.Status.NumberUnavailable == 0 {
+				log.Infof("Meshnet Healthy")
+				return nil
+			}
+		}
+	}
+}
+
+type Cluster interface {
+	Deploy(context.Context) error
+}
+
+type Ingress interface {
+	Deploy(context.Context) error
+	SetKClient(kubernetes.Interface)
+	Healthy(context.Context) error
+}
+
+type CNI interface {
+	Deploy(context.Context) error
+	SetKClient(kubernetes.Interface)
+	Healthy(context.Context) error
+}
+
+type Deployment struct {
+	Cluster Cluster
+	Ingress Ingress
+	CNI     CNI
+}
+
+type DeploymentConfig struct {
+	Cluster ClusterSpec `yaml:"cluster"`
+	Ingress IngressSpec `yaml:"ingress"`
+	CNI     CNISpec     `yaml:"cni"`
+}
+
+func NewDeployment(cfg *DeploymentConfig) (*Deployment, error) {
+	d := &Deployment{}
+	switch cfg.Cluster.Kind {
+	case "Kind":
+		log.Infof("Using kind scenario")
+		v := &KindSpec{}
+		if err := cfg.Cluster.Spec.Decode(v); err != nil {
+			return nil, err
+		}
+		d.Cluster = v
+	default:
+		return nil, fmt.Errorf("cluster type not supported: %s", cfg.Cluster.Kind)
+	}
+	switch cfg.CNI.Kind {
+	case "Meshnet":
+		v := &MeshnetSpec{}
+		if err := cfg.CNI.Spec.Decode(v); err != nil {
+			return nil, err
+		}
+		d.CNI = v
+	default:
+		return nil, fmt.Errorf("CNI type not supported: %s", cfg.CNI.Kind)
+	}
+	switch cfg.Ingress.Kind {
+	case "MetalLB":
+		v := &MetalLBSpec{}
+		if err := cfg.Ingress.Spec.Decode(v); err != nil {
+			return nil, err
+		}
+		d.Ingress = v
+	default:
+		return nil, fmt.Errorf("ingress type not supported: %s", cfg.Ingress.Kind)
+	}
+	return d, nil
+}
+
+var (
+	deploymentBasePath string
+)
+
+func deployFn(cmd *cobra.Command, args []string) error {
+	if len(args) != 1 {
+		return fmt.Errorf("%s: missing args", cmd.Use)
+	}
+	var err error
+	deploymentBasePath, err = filepath.Abs(args[0])
+	if err != nil {
+		return err
+	}
+	deploymentBasePath = filepath.Dir(deploymentBasePath)
+	b, err := ioutil.ReadFile(args[0])
+	if err != nil {
+		return err
+	}
+	dCfg := &DeploymentConfig{}
+	if err := yaml.Unmarshal(b, dCfg); err != nil {
```

Comment:
In yamlv2 there is a strict unmarshalling to catch yaml errs early?

https://pkg.go.dev/gopkg.in/yaml.v2#UnmarshalStrict

It seems in v3 there is no such thing?

Reply:
decoder := yaml.NewDecoder(bytes.NewBuffer(b))
	decoder.KnownFields(true)